src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,151 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        final Node<T> oldTail = tail;
+        final Node<T> newNode = new Node<>(oldTail, value, null);
+        tail = newNode;
+        if (oldTail == null) {
+            head = newNode;
+        } else {
+            oldTail.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (size == index) {
+            add(value);
+        } else {
+            Node<T> nextNode = node(index);
+            Node<T> previousNode = nextNode.prev;
+            Node<T> newNode = new Node<>(previousNode, value, nextNode);
+            if (previousNode == null) {
+                head = newNode;
+            } else {
+                previousNode.next = newNode;
+            }
+            if (nextNode == null) {
+                tail = newNode;
+            } else {
+                nextNode.prev = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> nodeOnIndex = node(index);
+        T oldvalue = nodeOnIndex.item;
+        nodeOnIndex.item = value;
+        return oldvalue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeOnIndex = node(index);
+        T removedValue = nodeOnIndex.item;
+        unlinkNode(nodeOnIndex);
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> temp = head;
+        int i = 0;
+        while (temp != null) {
+            if (temp.item == null ? object == null : temp.item.equals(object)) {
+                unlinkNode(temp);
+                return true;
+            }
+            temp = temp.next;
+            i++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(
+                    "this index doesn't exists in linked list: " + index);
+        }
+    }
+
+    private Node<T> node(int index) {
```

Comment:
bad method name, unclear what does it do
```suggestion
    private Node<T> getNodeByIndex(int index) {
```